Mathlib/GroupTheory/Perm/Sign.lean

```diff
@@ -112,14 +113,34 @@ theorem swap_induction_on [Finite α] {P : Perm α → Prop} (f : Perm α) :
         (ih _ ⟨rfl, fun v hv => hl.2 _ (List.mem_cons_of_mem _ hv)⟩ h1 hmul_swap)
 #align equiv.perm.swap_induction_on Equiv.Perm.swap_induction_on
 
-theorem closure_isSwap [Finite α] : Subgroup.closure { σ : Perm α | IsSwap σ } = ⊤ := by
+theorem mclosure_isSwap [Finite α] : Submonoid.closure { σ : Perm α | IsSwap σ } = ⊤ := by
   cases nonempty_fintype α
-  refine eq_top_iff.mpr fun x _ => ?_
+  refine top_unique fun x _ ↦ ?_
   obtain ⟨h1, h2⟩ := Subtype.mem (truncSwapFactors x).out
   rw [← h1]
-  exact Subgroup.list_prod_mem _ fun y hy => Subgroup.subset_closure (h2 y hy)
+  exact Submonoid.list_prod_mem _ fun y hy ↦ Submonoid.subset_closure (h2 y hy)
+
+theorem closure_isSwap [Finite α] : Subgroup.closure { σ : Perm α | IsSwap σ } = ⊤ :=
+  Subgroup.closure_eq_top_of_mclosure_eq_top mclosure_isSwap
 #align equiv.perm.closure_is_swap Equiv.Perm.closure_isSwap
 
+/-- Every finite symmetric group is generated by transpositions of adjacent elements. -/
+theorem mclosure_swap_castSucc_succ (n : ℕ) :
+    Submonoid.closure (Set.range fun i : Fin n ↦ swap i.castSucc i.succ) = ⊤ :=
+  top_unique <| by
+    rw [← mclosure_isSwap, Submonoid.closure_le]
+    rintro _ ⟨i, j, ne, rfl⟩
+    wlog lt : i < j generalizing i j
+    · rw [swap_comm]; exact this _ _ ne.symm (ne.lt_or_lt.resolve_left lt)
+    clear ne; induction' j using Fin.induction with j ih
+    · cases lt
+    have mem : ?_ := ?_
```

Comment:
That's really hard to read (although extremely efficient!). Can you come up with something more explicit?

Reply:
The whole type is simply too long and doesn't even fit in one line, so I added a comment
`-- swap i j.succ ∈ Submonoid.closure ... (the LHS)`

Reply:
I still think this is not good style, because when you read the proof `mem` is not assigned (even its type), so there is no way the reader can understand what is going on even using the goal view. Even if the type takes two or three lines, you should just write it down to help your reader.

Reply:
Okay thanks. I think I copied the goal rather than the type of `mem` 🤦